### PR TITLE
feat:Use of data handlers for Vision LM support 

### PR DESCRIPTION
**Pull request by Abhishek-TAMU**
### Description of the change

Update the code for VLM support to use data handlers to process dataset before using `VisionDataCollator`.
Changes:

- Use of Data handlers to give user ability to specify `chat_template` and other `processor` flag.
- Restructure `VisionDataCollator` to make a batch and perform masking on labels.

Changes contribution to PR: https://github.com/foundation-model-stack/fms-hf-tuning/pull/451

### Related issue number

Issue: https://github.ibm.com/ai-foundation/watson-fm-stack-tracker/issues/1610

### Was the PR tested

- [ ] I have added >=1 unit test(s) for every new method I have added.
- [ ] I have ensured all unit tests pass

### CODE REVIEW

**Comment by Abhishek-TAMU** [comment]
Testing in progress: Testing of vision dataset with data handlers is ongoing. Would post updates shortly. 
Feel free to update repo `anhuong:vision-model` with latest code from main before reviewing this PR. @anhuong

**Comment by Abhishek-TAMU** [comment]
Tested this PR and Data Preprocessor part is functional now and ready for E2E testing. @anhuong

**Comment by Abhishek-TAMU** [comment]
Now testing the `VisionCollator` by e2e testing of `Llama3.2-11B-Vision-Instruct`.

**Comment by Abhishek-TAMU** [comment]
Ran e2e test successfully for LoRA tuning with below config:

```
{
  "model_name_or_path": "meta-llama/Llama-3.2-11B-Vision-Instruct",
  "training_data_path": "HuggingFaceH4/llava-instruct-mix-vsft",
  "output_dir": "/data/abhishek/vlm_work/fms-hf-tuning/output/llama_32_vision",
  "save_model_dir": "/data/abhishek/vlm_work/fms-hf-tuning/output/llama_32_vision/save_model",
  "num_train_epochs": 3.0,
  "per_device_train_batch_size": 2,
  "gradient_accumulation_steps": 1,
  "learning_rate": 1e-4,
  "response_template": "\n### Response:",
  "dataset_text_field": "output",
   "bf16": true,
  "torch_dtype": "bfloat16",
  "use_flash_attn": false,
  "remove_unused_columns": false,
  "dataset_kwargs": {"skip_prepare_dataset": true},
  "multimodal": true,
  "peft_method": "lora",
  "r": 8,
  "lora_dropout": 0.05,
  "lora_alpha": 16,
  "target_modules": ["all-linear"],
  "lora_post_process_for_vllm": true,
  "gradient_checkpointing": true,
  "text_field_name": "messages",
  "image_field_name": "images"
}
```

File: tuning/data/data_handlers.py [hunk 1]
```diff
@@ -247,16 +247,48 @@ def apply_tokenizer_chat_template(
...
         Saves the result to dataset_text_field argument.
     """
+    chat_template_key = kwargs.get("chat_template_key", None)
     if tokenizer.chat_template is None:
         raise ValueError(
             "Tokenizer does not contain tokenizer.chat_template\
                           please pass data_args.chat_template"
         )
+    if chat_template_key and chat_template_key in element:
+        element = element[chat_template_key]
```
Comment (anhuong) [inline tuning/data/data_handlers.py:257]: can you explain what the chat_template_key is doing? why does it need to be passed to this data_handler in order to get the text value and how does this data_handler work for applying chat templates to other dataset?

File: tuning/data/setup_dataprocessor.py [hunk 2]
```diff
@@ -191,6 +191,52 @@ def _get_default_dataset_handlers(data_args, tokenizer_kwargs):
...
+        "dataset_text_field": data_args.text_field_name,
+        "chat_template_key": data_args.text_field_name,
+    }
+    kwargs1 = {
+        "fn_kwargs": fn_kwargs1,
+        "batched": False,
+        "remove_columns": None,
+    }
+    handlers.append(
+        DataHandlerConfig("apply_tokenizer_chat_template", arguments=kwargs1)
```
Comment (anhuong) [inline tuning/data/setup_dataprocessor.py:211]: both data handlers are manipulating the same dataset one after the other correct? so first the chat_template is applied and then the processor that combines the image and text is applied

**Review by anhuong — COMMENTED** [review]
few questions for clarity but otherwise the changes look good to me

File: tuning/data/data_handlers.py (hunk 1 above)
Comment (Abhishek-TAMU) [reply tuning/data/data_handlers.py:257]: Sure. As per earlier code [here](https://github.com/foundation-model-stack/fms-hf-tuning/blob/87d987a6f017d5c0254b4f42c0f5b2aef459e1fb/tuning/data/data_handlers.py#L255) this handler just applies `chat_template` where `chat_template` is like this in our unit test case [here](https://github.com/foundation-model-stack/fms-hf-tuning/blob/87d987a6f017d5c0254b4f42c0f5b2aef459e1fb/tests/test_sft_trainer.py#L947) where key/text_field of element `messages` is already taken from data and processed. 

But for chat templates of `Llama-3.2-11b-Vision-Instruct` [here](https://huggingface.co/meta-llama/Llama-3.2-11B-Vision-Instruct/blob/main/tokenizer_config.json#L2061) and for `llava-hf/llava-v1.6-mistral-7b-hf` [here](https://huggingface.co/llava-hf/llava-v1.6-mistral-7b-hf/blob/main/tokenizer_config.json#L49) we can see that automatically `"messages"` field is not being read and hence we have to pass `chat_template_key="messages"` OR any `text_field` of the dataset which should be applied on with `chat_template` and ignore all other fields such as `image_field`.

File: tuning/data/setup_dataprocessor.py (hunk 2 above)
Comment (Abhishek-TAMU) [reply tuning/data/setup_dataprocessor.py:211]: Yes that's correct.

**Comment by Abhishek-TAMU** [comment]
Just pushed minor error handling code as an improvement. @anhuong 
Feel free to merge or discuss on any part of the code.

File: tuning/data/data_handlers.py (hunk 1 above)
Comment (anhuong) [reply tuning/data/data_handlers.py:257]: > how does this data_handler work for applying chat templates to other dataset

I was asking this because why don't other models require passing in the chat_template_key to apply the chat_template? The vision chat templates also start with `{% for message in messages %}`, but you're saying that for other models this is already referring to the correct column in the dataset but for vision model it does not?

And for other uses of this data handler, the dataset_text_field is only for the newly formatted output key whereas for the vision model we will be using dataset_text_field also for the column in the dataset that contains the text

Comment (Abhishek-TAMU) [reply tuning/data/data_handlers.py:257]: I am not sure about how other models are using this handler (Dushyant might know more on what models or what teams are using this handler). I made change of adding `chat_template_key` based on this [unit test case](https://github.com/foundation-model-stack/fms-hf-tuning/blob/87d987a6f017d5c0254b4f42c0f5b2aef459e1fb/tests/test_sft_trainer.py#L947) which Dushyant had added where chat template starts with `{% for message in messages['messages'] %}` compared to chat template in vision model `{% for message in messages %}` and hence I added a separate case if user passes `chat_template_key` then use that OR else the default case will be applied.

Comment (Abhishek-TAMU) [reply tuning/data/data_handlers.py:257]: > And for other uses of this data handler, the dataset_text_field is only for the newly formatted output key whereas for the vision model we will be using dataset_text_field also for the column in the dataset that contains the text.

My idea was to use same `text_field_name` (`messages` in this case) to stored chat template processed text by replacing normal text, and then further use this `text_field_name` when applying `processor` in second handler. 

As normal text is not used anywhere, I used same key for `dataset_text_field`. Other option is for some reason if we prefer to keep the original text, then we can also choose some other `dataset_text_field` which stores processed text and use this `dataset_text_field` as `text_field_name` for second handler.

Comment (anhuong) [reply tuning/data/data_handlers.py:257]: I plan to refactor my changes to re-use dataset_text_field instead of text_field_name as per Dushyant's comment in my PR which I agree with, better to reuse the field name. So then for image+text datasets it will require dataset_text_field and dataset_image_field for the column names. So i thin that aligns with your first option here

**Review by anhuong — APPROVED** [review]

File: tuning/data/data_handlers.py (hunk 1 above)
Comment (Abhishek-TAMU) [reply tuning/data/data_handlers.py:257]: This also sounds good. As we are already tokenizing `dataset_text_field` and `dataset_image_field` to generate `input_ids, attention_mask, pixel_values, etc.`, hence SFTTrainer probably would ignore `dataset_text_field` text field passed to it and just use `input_ids` etc...
